java/sage/io/RemoteSageFile.java

```diff
@@ -0,0 +1,555 @@
+      outStream.flush();
+      returnValue = inStream.read();
+    }
+
+    remoteOffset += 1;
+    return returnValue & 0xFF;
+  }
+
+  @Override
+  public int read(byte[] b) throws IOException
+  {
+    return read(b, 0, b.length);
+  }
+
+  @Override
+  public synchronized int read(byte[] b, int off, int len) throws IOException
+  {
+    if (len == 0)
+      return len;
+
+    // This cannot return anything greater than len.
+    len = (int)getMaxRead(remoteOffset, len);
+
+    // End of the file.
+    if (len == 0)
+      return -1;
+
+    try
+    {
+      outStream.write(("READ " + remoteOffset + " " + len + "\r\n").getBytes(Sage.BYTE_CHARSET));
+      outStream.flush();
+      inStream.readFully(b, off, len);
+    }
+    catch (IOException e)
+    {
+      reconnect();
+      outStream.write(("READ " + remoteOffset + " " + len + "\r\n").getBytes(Sage.BYTE_CHARSET));
+      outStream.flush();
+      inStream.readFully(b, off, len);
+    }
+
+    remoteOffset += len;
+    return len;
+  }
+
+  @Override
+  public void readFully(byte[] b) throws IOException
+  {
+    readFully(b, 0, b.length);
+  }
+
+  @Override
+  public void readFully(byte[] b, int off, int len) throws IOException
+  {
+    if (len == 0)
+      return;
+
+    int bytesRead = read(b, off, len);
+
+    while (len > 0)
+    {
+      if (bytesRead == -1)
+        throw new EOFException("End of file reached!");
+
+      len -= bytesRead;
+      off += bytesRead;
+
+      if (len == 0)
+        break;
+
+      bytesRead = read(b, off, len);
+    }
+  }
+
+  @Override
+  public synchronized void write(int b) throws IOException
+  {
+    if (readonly)
+      throw new IOException("Remote file is read only.");
+
+    try
+    {
+      outStream.write(("WRITE " + remoteOffset + " 1\r\n").getBytes(Sage.BYTE_CHARSET));
+      outStream.flush();
+      outStream.write(b);
+    }
+    catch (IOException e)
+    {
+      reconnect();
+      outStream.write(("WRITE " + remoteOffset + " 1\r\n").getBytes(Sage.BYTE_CHARSET));
+      outStream.flush();
+      outStream.write(b);
+    }
+
+    remoteOffset += 1;
+  }
+
+  @Override
+  public void write(byte[] b) throws IOException
+  {
+    write(b, 0, b.length);
+  }
+
+  @Override
+  public synchronized void write(byte[] b, int off, int len) throws IOException
+  {
+    if (readonly)
+      throw new IOException("Remote file is read only.");
+
+    if (len == 0)
+      return;
+
+    try
+    {
+      outStream.write(("WRITE " + remoteOffset + " " + len + "\r\n").getBytes(Sage.BYTE_CHARSET));
+      outStream.flush();
+      outStream.write(b, off, len);
+    }
+    catch (IOException e)
+    {
+      reconnect();
+      outStream.write(("WRITE " + remoteOffset + " " + len + "\r\n").getBytes(Sage.BYTE_CHARSET));
+      outStream.flush();
+      outStream.write(b, off, len);
+    }
+
+    remoteOffset += len;
+  }
+
+  @Override
+  public void randomWrite(long pos, byte[] b, int off, int len) throws IOException
```

Comment:
This should be synchronized since you are changing remoteOffset in this method which can be changed in our synchronized methods. (of course you're totally asking for trouble if you're using a file stream that way anyways...but hey...correctness and all :) )